lib/pdk/tests/unit.rb

```diff
@@ -6,20 +6,89 @@ module PDK
+        # TODO: test selection
+        [File.join(PDK::Util.module_root, 'bin', 'rake'), 'spec']
       end
 
-      def self.invoke(tests, report = nil)
+      def self.invoke(report, options = {})
         PDK::Util::Bundler.ensure_bundle!
+        PDK::Util::Bundler.ensure_binstubs!('rake')
 
-        puts _('Running unit tests: %{tests}') % { tests: tests }
+        tests = options.fetch(:tests)
 
-        output = PDK::CLI::Exec.execute(cmd(tests))
-        report.write(output) if report
+        cmd_argv = cmd(tests)
+        cmd_argv.unshift('ruby') if Gem.win_platform?
+
+        command = PDK::CLI::Exec::Command.new(*cmd_argv).tap do |c|
+          c.context = :module
+          c.add_spinner('Running unit tests')
+          c.environment['CI_SPEC_OPTIONS'] = '--format j'
+        end
+
+        PDK.logger.debug(_('Running %{cmd}') % { cmd: command.argv.join(' ') })
+
+        result = command.execute!
+
+        # TODO: cleanup rspec and/or beaker output
+        # Iterate through possible JSON documents until we find one that is valid.
+        json_result = nil
+
+        result[:stdout].scan(%r{\{(?:[^{}]|(?:\g<0>))*\}}x) do |str|
+          begin
+            json_result = JSON.parse(str)
+            break
+          rescue JSON::ParserError
+            next
+          end
+        end
+
+        raise PDK::CLI::FatalError, _('Unit test output did not contain a valid JSON result: %{output}') % { output: result[:stdout] } unless json_result
+
+        parse_output(report, json_result)
+
+        result[:exit_code]
+      end
+
+      def self.parse_output(report, json_data)
+        if json_data['messages'] && !json_data['messages'].empty?
+          json_data['messages'].each do |msg|
+            report.add_event(
+              source: 'rspec',
+              state: :error,
+              message: msg,
+            )
+          end
+        elsif json_data['summary'] && json_data['summary']['failure_count'].zero?
+          report.add_event(state: :passed, severity: :ok)
+        elsif json_data['examples']
+          failures = json_data['examples'].reject { |e| e['status'] == 'passed' }
+
+          failures.each do |f|
+            report.add_event(
+              source: 'rspec',
+              state: :failure,
+              file: f['file_path'],
+              line: f['line_number'],
+              test: f['full_description'],
+              severity: f['status'],
+              message: f['exception']['message'].gsub(%r{\s+}, ' '),
+              trace: f['exception']['backtrace'],
+            )
+          end
+        else
+          raise PDK::CLI::FatalError, _('Unable to parse results of unit test tool invocation: %{results}') % { results: json_data.to_s }
+        end
+
+        return unless json_data['summary']
+
+        # TODO: standardize summary output
+        $stderr.puts '  ' << _('Evaluated %{total} tests in %{duration} seconds: %{errors} errors, %{failures} failures, %{pending} pending') % {
```

Comment:
(Note also that a report generator could create it's own summary when appropriate based on the events we added when parsing the output.)